src/cdk/tree/tree.ts

```diff
@@ -171,8 +331,34 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
     }
   }
 
-  // TODO(tinayuangao): Work on keyboard traversal and actions, make sure it's working for RTL
-  //     and nested trees.
+  /**
+   * Sets the node type for the tree, if it hasn't been set yet.
+   *
+   * This will be called by the first node that's rendered in order for the tree
+   * to determine what data transformations are required.
+   */
+  _setNodeTypeIfUnset(nodeType: 'flat' | 'nested') {
+    if (this._nodeType.value === null) {
+      this._nodeType.next(nodeType);
+    }
+  }
+
+  /**
+   * Sets the tabIndex on the host element.
+   *
+   * NB: we don't set this as a host binding since children being activated
+   * (e.g. on user click) doesn't trigger this component's change detection.
+   */
+  _setTabIndex() {
+    // If the `TreeKeyManager` has no active item, then we know that we need to focus the initial
+    // item when the tree is focused. We set the tabindex to be `0` so that we can capture
+    // the focus event and redirect it. Otherwise, we unset it.
+    if (!this._keyManager.getActiveItem()) {
+      this._elementRef.nativeElement.setAttribute('tabindex', '0');
+    } else {
+      this._elementRef.nativeElement.removeAttribute('tabindex');
+    }
```

Comment:
As a general point, direct DOM manipulation like this should now always be in `afterRender` or `afterNextRender`

In this case, though, should this not be set via binding?

Reply:
I'm working on addressing this by seeing if we can avoid setting tabindex on the `role="tree"` element.

Reply:
We fixed this by `focus()` and `unfocus()` method on each tree node. We removed the tabindex attribute on the role="tree" element.